makeflow/src/makeflow_wrapper_enforcement.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (C) 2015- The University of Notre Dame
+ * This software is distributed under the GNU General Public License.
+ * See the file COPYING for details.
+ * */
+
+#include "stringtools.h"
+#include "xxmalloc.h"
+#include "copy_stream.h"
+#include "debug.h"
+
+#include "list.h"
+#include "dag.h"
+#include "makeflow_wrapper.h"
+#include "makeflow_wrapper_enforcement.h"
+
+#include <string.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <errno.h>
+
+#define enforcer_pattern "enforcer_"
+#define mountlist_pattern "mount_"
+#define tmp_pattern "tmp_"
+#define vartmp_pattern "vartmp_"
+#define shm_pattern "shm_"
+#define local_parrot_path "parrot_run"
+
+void makeflow_log_file_expectation( struct dag *d, struct list *file_list );
+void makeflow_log_file_existence( struct dag *d, struct list *file_list );
+
+void makeflow_make_tempdir(const char* path) {
+	struct stat buf;
+	if (mkdir(path, S_IRWXU)) {
+		if ((errno != EEXIST) || stat(path, &buf) || !S_ISDIR(buf.st_mode) ) {
+			fatal("could not create temp directory: %s", strerror(errno));
+		}
+	}
+}
+
+void makeflow_wrapper_enforcer_init(struct makeflow_wrapper *w, char *parrot_path) {
+	struct stat stat_buf;
+	int host_parrot = open(parrot_path, O_RDONLY);
+	if (host_parrot == -1) {
+		fatal("could not open parrot at `%s': %s", parrot_path, strerror(errno));
+	}
+	fstat(host_parrot, &stat_buf);
+	if (!(stat_buf.st_mode & (S_IXUSR|S_IXGRP|S_IXOTH))) {
+		fatal("%s is not executable", parrot_path);
+	}
+	int local_parrot = open(local_parrot_path, O_WRONLY|O_CREAT, S_IRWXU);
+	if (local_parrot == -1) {
+		fatal("could not create local copy of parrot: %s", strerror(errno));
+	} else {
+		fchmod(local_parrot, 0755);
+		if (copy_fd_to_fd(host_parrot, local_parrot) != stat_buf.st_size) {
+			fatal("could not copy parrot: %s");
+		}
+	}
+	close(local_parrot);
+	close(host_parrot);
+
+	makeflow_wrapper_add_input_file(w, local_parrot_path);
+	makeflow_wrapper_add_input_file(w, xxstrdup(enforcer_pattern "%%"));
+	makeflow_wrapper_add_input_file(w, xxstrdup(mountlist_pattern "%%"));
+	makeflow_wrapper_add_input_file(w, xxstrdup(tmp_pattern "%%"));
```

Comment:
These input files no longer need to be specified as they are created at the worker and deleted there as well. Removing line 68-70 fixed an issue with WQ failures.